### PR TITLE
[NG] Tree view: Eager declarative case

### PR DESCRIPTION
**Pull request by youdz**
This case is the simple one: the nodes are declared in HTML on the app's side, and we know all of them from the start so we can be smart about selection.

As previously advertised, unit tests are not exhaustive because of time constraints, but they still offer reasonable coverage.

I apologize for this PR being much bigger than I'd normally like, but deciding to split by feature means the overhead on the first PR is real. One thing that can help you out is that since **I did not remove the old SCSS file**, I kept the tree node's template pretty much the same. Because it was such a rewrite, I went with copy-pasting the small pieces of old code that we could reuse, rather than edit in place and generate obfuscated diffs. The existing code that you can quickly skim through are:
- The node's template (minimal changes)
- ARIA attributes and roles (no change)
- Animations (no change)
- Tree with icons demo
- Routing tree demo

I promise upcoming PRs will be much smaller.

### CODE REVIEW

File: src/clr-angular/data/tree-view/tree-node.ts [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+  SkipSelf,
+} from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { Expand } from '../../utils/expand/providers/expand';
+import { ClrCommonStrings } from '../../utils/i18n/common-strings.interface';
+import { UNIQUE_ID, UNIQUE_ID_PROVIDER } from '../../utils/id-generator/id-generator.service';
+import { LoadingListener } from '../../utils/loading/loading-listener';
+import { DeclarativeTreeNodeModel } from './models/declarative-tree-node.model';
+import { ClrSelectedState } from './models/selected-state';
```
Comment (gnomeontherun) [inline src/clr-angular/data/tree-view/tree-node.ts]: Missed this before, but we should add `.enum` to filename

File: src/dev/src/app/tree-view/tree-node-routing/tree-node-routing-rubber-soul.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'clr-tree-node-routing-rubber-soul-demo',
```
[the remark this answers is not in the document]
Comment (youdz) [reply src/dev/src/app/tree-view/tree-node-routing/tree-node-routing-rubber-soul.ts]: We actually had a bug a while back with routing components without selectors, but I don't remember what it was. It created styling issues.

File: src/dev/src/app/tree-view/nodes-with-icons/nodes-with-icons.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import { Component } from '@angular/core';
+
+// Import icon sets for this demo
+import '@clr/icons/shapes/chart-shapes';
+import '@clr/icons/shapes/essential-shapes';
+import '@clr/icons/shapes/technology-shapes';
+import '@clr/icons/shapes/travel-shapes';
+import '@clr/icons/shapes/social-shapes';
```
[the remark this answers is not in the document]
Comment (youdz) [reply src/dev/src/app/tree-view/nodes-with-icons/nodes-with-icons.ts]: Haha, I just copy-pasted this demo from the old tree. I'll trim them to stick to core, no worries. 👍

File: src/clr-angular/utils/expand/if-expanded.ts [hunk 4]
```diff
@@ -57,16 +66,29 @@ export class ClrIfExpanded implements OnInit, OnDestroy {
...
     } else {
-      // TODO: Move when we move the animation logic to Datagrid Row Expand
-      // We clear before the animation is over. Not ideal, but doing better would involve a much heavier
-      // process for very little gain. Once Angular animations are dynamic enough, we should be able to
-      // get the optimal behavior.
-      this.container.clear();
+      try {
+        // If we don't have a template ref, we fallback to a crude display: none for now.
+        if (this.expand.expanded) {
+          this.renderer.setStyle(this.el.nativeElement, 'display', null);
```
[the remark this answers is not in the document]
Comment (youdz) [reply src/clr-angular/utils/expand/if-expanded.ts]: No, using `ElementRef.nativeElement` with the renderer works for SSR and webworkers. That's the whole point of the renderer.
We need guards for `ElementRef.nativeElement.xxx` access to actual native element properties.

**Pull request merged**